generators/generator-base-private.js

```diff
@@ -1214,14 +1205,10 @@ module.exports = class JHipsterBasePrivateGenerator extends Generator {
       throw new Error(`option 'hostname' is required for ${databaseType} databaseType`);
     }
     let dbcUrl;
```

Comment:
What about extract to `sql-constants.js`?
```suggestion
    const { databaseData } = require('../sql-constants.js');
    let databaseDataForType = databaseData[databaseType];
    if (!databaseDataForType) {
      throw new Error(`${databaseType} databaseType is not supported`);
    }
    if (databaseDataForType[protocol]) {
      databaseDataForType = {
        ...databaseDataForType,
        ...databaseDataForType[protocol],
      };
    }
    const { protocolSuffix = '', extraOptions = '', useDirectory = false } = databaseDataForType;
    let { port = '' } = databaseDataForType;
    if (useDirectory && !options.localDirectory) {
      throw new Error(`'localDirectory' option should be provided for ${databaseType} databaseType`);
    }
    if (options.itests && H2_MEMORY === databaseType) {
      port = ':12344';
    }
    return `${protocol}:${protocolSuffix}${options.hostname}${port}${useDirectory ? options.localDirectory : options.databaseName}${options.skipExtraOptions ? '' : extraOptions}`;
```

```js
const databaseData = {
  [MSSQL]: {
    protocolSuffix: 'sqlserver://',
    port: ':1433/',
    r2dbc: {
      protocolSuffix: 'mssql://',
      port: ':1433;database=',
    },
  },
  [MYSQL]: {
    protocolSuffix: 'mysql://',
    port: ':3306/',
    extraOptions: '?useUnicode=true&characterEncoding=utf8&useSSL=false&useLegacyDatetimeCode=false&serverTimezone=UTC&createDatabaseIfNotExist=true',
  },
  [MARIADB]: {
    protocolSuffix: 'mariadb://',
    port: ':3306/',
    extraOptions: '?useLegacyDatetimeCode=false&serverTimezone=UTC',
  },
  [POSTGRESQL]: {
    protocolSuffix: 'postgresql://',
    port: ':5432/',
    extraOptions: '?useLegacyDatetimeCode=false&serverTimezone=UTC',
  },
  [POSTGRESQL]: {
    protocolSuffix: 'postgresql://',
    port: ':5432/',
    extraOptions: '?useLegacyDatetimeCode=false&serverTimezone=UTC',
  },
  [ORACLE]: {
    protocolSuffix: 'oracle:thin:@',
    port: ':1521:',
  },
  [H2_DISK]: {
    protocolSuffix: 'h2:file:///',
    port: ':1521:',
    useDirectory: true,
    extraOptions: ';DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE',
    r2dbc: {
      protocolSuffix: 'h2:file:',
    },
  },
  [H2_MEMORY]: {
    protocolSuffix: 'h2:mem:///',
    port: ':1521:',
    extraOptions: ';DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE',
    r2dbc: {
      protocolSuffix: 'h2:mem:',
    },
  },
};
```

Reply:
This notation is like a foreign language to me: I'm a former Java guy, not a js one.
Nevertheless, I get the idea and it's pretty cool: can you implement it on a specific branch?
Otherwise I'll do it in a second step after this merge (it's extra to me)

Reply:
https://github.com/jhipster/generator-jhipster/issues/18759

Reply:
see https://github.com/Tcharl/generator-jhipster/tree/testcontainers_sql_rmshima